src/main/java/duke/ui/Ui.java

```diff
@@ -0,0 +1,154 @@
+        }
+        return composeBlock(text);
+    }
+
+    /**
+     * Returns a string representation of a message that the task has been marked as done.
+     * @param task the task that has been marked as done.
+     * @return a string representation of a message that the task has been marked as done.
+     */
+    public String showDoneTask(Task task) {
+        return composeBlock("Nice! I've marked this task as done:", "  " + task);
+    }
+
+    /**
+     * Returns a string representation of a message that the task has been removed from the list.
+     * @param task the task that has been removed.
+     * @param taskListSize the number of tasks in the task list.
+     * @return a string representation of a message that the task has been removed from the list.
+     */
+    public String showRemovedTask(Task task, int taskListSize) {
+        return composeBlock("Noted. I've removed this task:", "  " + task,
+                String.format("Now you have %d task%s in the list.",
+                        taskListSize, taskListSize > 1 ? "s" : ""));
+    }
+
+    /**
+     * Returns a string representation of the message that the task has been added to the list.
+     * @param task the task that has been added.
+     * @param taskListSize the number of tasks in the task list.
+     * @return a string representation of the message that the task has been added to the list.
+     */
+    public String showAddedTask(Task task, int taskListSize) {
+        return composeBlock("Got it. I've added this task:", "  " + task,
+                String.format("Now you have %d task%s in the list.",
+                        taskListSize, taskListSize > 1 ? "s" : ""));
+    }
+
+    /**
+     * Returns a string representation of all tasks that matches a keyword typed by the user.
+     * @param taskList a list tasks that matches a keyword.
+     * @return a string representation of all tasks that matches a keyword typed by the user.
+     */
+    public String showMatchingTasks(Task[] taskList) {
+        String[] text = new String[taskList.length + 1];
+        text[0] = "Here are the matching tasks in your list:";
+        for (int i = 0; i < taskList.length; i++) {
+            text[i + 1] = (i + 1) + "." + taskList[i];
+        }
+        return composeBlock(text);
+    }
+
+    /**
+     * Returns a string representation of saying good bye to the user.
+     * @return a string representation of saying good bye to the user.
+     */
+    public String sayGoodBye() {
+        return composeBlock("Bye. Hope to see you again soon!");
+    }
+
+    /**
+     * Returns a string representation of errors that occurs when parsing commands.
+     * @param exception the exception thrown when parsing commands.
+     * @return a string representation of errors that occurs when parsing commands.
+     */
+    public String showParsingError(Exception exception) {
+        return composeBlock("OPPS!!! " + exception.getMessage());
```

Comment:
Perhaps you may want to consider returning the stack trace here (possible using a combination of StringWriter and PrintWriter)? There are two exceptions found that may (subjectively) not be displaying sufficient information to the user as follows:

Case 1: When file path is not found
Methods such as "list", "todo" all do not work because when Duke.java calls the execute command, "tasks" attribute stored in Duke is empty (since loading fails and the TaskList is never initialised). A NullPointerException is returned, but when the user enters a command, the message shown is "OPPS!!! null"

Case 2: When an invalid time is taken in as input
eg. "event testEvent /at 2200-2300" is entered, the program shows "OPPS!! For input string: “00-2300” to the user because a NumberFormatException was thrown.

Alternatively, these two exceptions could be caught such that a more descriptive message is shown to the user.

These are just some suggestions after some testing; feel free to let me know if you happen to have any alternative points of view so I can learn if you happen to have any other ways of approaching this too, or if it should be left as it is!